### PR TITLE
osd: propagate pg log errors to cache tier

### PR DESCRIPTION
**Pull request by dillaman**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/osd/PGLog.h [hunk 1]
```diff
@@ -289,13 +289,20 @@ struct PGLog : DoutPrefixProvider {
...
+	uint32_t idx = 0;
 	for (auto i = p->second->extra_reqids.begin();
 	     i != p->second->extra_reqids.end();
-	     ++i) {
+	     ++idx, ++i) {
 	  if (i->first == r) {
 	    *version = p->second->version;
 	    *user_version = i->second;
 	    *return_code = p->second->return_code;
+	    if (*return_code >= 0) {
```
Comment (liewegas) [inline src/osd/PGLog.h]: maybe == 0 instead of >= 0?  I'm not sure we ever store a >0 value, but if we did, it seems like we might want to return it?

Comment (liewegas) [reply src/osd/PGLog.h]: @jdurgin or maybe we should prefer negative error codes here?

Comment (liewegas) [reply src/osd/PGLog.h]: I guess it's a moot point since this is always going to be a promote log entry, where we know the return_code is always 0

Comment (dillaman) [reply src/osd/PGLog.h]: > return_code is always 0

OK, I didn't know if that was true or not and I was trying to allow any cache tier PG log failures supersede the base tier return codes.

**Comment by liewegas** [comment]
I don't suppose we can add add a test case to tier.cc?

**Comment by dillaman** [comment]
> I don't suppose we can add add a test case to tier.cc?

Shouldn't be an issue -- just need to issue a guarded write op that returns a failure when the expectation is not met. The current RBD CLI tests hit this bug on every run when the cache tier is enabled.

**Comment by liewegas** [comment]
a test would be nice!

**Comment by dillaman** [comment]
> a test would be nice!

Ask and you shall receive -- rebased and attached a reproducer test

**Comment by gregsfortytwo** [comment]
jenkins retest this please

**Comment by dillaman** [comment]
RBD test run was clean: http://pulpito.ceph.com/jdillaman-2018-10-25_20:39:02-rbd-wip-jd-testing-distro-basic-smithi/

**Comment by gregsfortytwo** [comment]
Should still go through RADOS run as well if anybody has one convenient, @neha-ojha, @tchaikov

**Pull request merged**